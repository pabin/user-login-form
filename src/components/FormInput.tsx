import React, { useState } from 'react';
import {
  EMAIL_VALIDATION_MESSAGE,
  PASSWORD_HELP_TEXT,
  EMAIL,
  PASSWORD,
} from '../constants/form';

import styles from './styles/formInput.module.css';
import eye from '../assets/icons/eye.png';

type FormInputType = {
  name: string,
  type: string,
  value: string,
  setValue: (value: string) => void,
  validationError?: boolean,
}

const FormInput  = ({ type, name, value, setValue, validationError }: FormInputType) => {
  const [isFocused, setFocused] = useState(false)
  const [viewPassword, setViewPassword] = useState(false)

  return (
    <div className={styles.container}>

      <fieldset className={validationError ? `${styles.fieldsetError} ${styles.fieldset}` : isFocused ? `${styles.fieldset} ${styles.fieldsetFocus}` : styles.fieldset}>
        <legend data-testid={`legend-${name}`} className={validationError ? `${styles.legendError} ${styles.legend}` : styles.legend} >
          { isFocused || value ? name : null }
        </legend>
        <input
          data-testid={name}
          type={viewPassword ? 'text'  : type}
          className={styles.input}
          placeholder={!isFocused ? name : ''}
          // value={value}
          onChange={(e) => setValue(e.target.value)}
          onFocus={() => setFocused(true)}
          onBlur={() => setFocused(false)}
        />
        { (type === PASSWORD) && (
          <img data-testid="eye" onClick={() => setViewPassword(prevState => !prevState)} className={styles.icon} src={eye} alt="eye" width="18" height="18" />
        )}
      </fieldset>
      { (type === EMAIL && validationError) && (
        <span data-testid="emailError" className={`${styles.helpText} ${styles.validationError}`}>{ EMAIL_VALIDATION_MESSAGE }</span>
      )}
      { type === PASSWORD && (
        <span data-testid="passwordHelpText" className={validationError ? `${styles.helpText} ${styles.validationError}` : styles.helpText}>{ PASSWORD_HELP_TEXT }</span>
      )}
    </div>
  )
}

export default FormInput;