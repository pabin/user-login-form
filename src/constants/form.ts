export const FORM_HEADING = "Let's set up your account";
export const ALREADY_HAVE_ACCOUNT_MSG = "Already have an account?";
export const SIGIN_IN = "Sign in";

// form fields
export const YOUR_NAME = "Your Name";
export const EMAIL_ADDRESS = "Email Address";
export const PASSWORD_LABEL = "Password";
export const BUTTON_TEXT = "Next";
export const SELECT_TITLE = "I would describe my user type as";
export const SELECT_OPTION1 = "Developer";
export const SELECT_OPTION2 = "QA Engineer";
export const SELECT_OPTION3 = "System Admin";

export const EMAIL_VALIDATION_MESSAGE = "Please enter a valid email address";
export const PASSWORD_HELP_TEXT = "Minimum 8 Characters";

export const TERMS_TEXT = 'Terms of Service';
export const PRIVACY_TEXT = 'Privacy Policy';
export const AGGREEMENT_TEXT = `By clicking the "Next" button, you agree to creating a free account, and to`;

// input types
export const EMAIL = 'email';
export const PASSWORD = 'password';
export const TEXT = 'text';

export const BUTTON_LABEL = 'Next';