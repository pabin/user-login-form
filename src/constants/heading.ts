export const HEADING_TITLE = 'Dummy Heading';
export const HEADING_DESCRIPTION = 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore at dolore magna aliqua';