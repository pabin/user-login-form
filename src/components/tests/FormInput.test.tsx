import React from 'react';
import { render, fireEvent, cleanup, screen } from '@testing-library/react'
import { PASSWORD_LABEL } from '../../constants/form';
import FormInput from '../FormInput';

describe('FormInput', () => {
  afterEach(cleanup);

  describe('when input type is password', () => {
    const props = {
      type: 'password',
      name: 'Password',
      value: 'randompassword',
      validationError: false,
      setValue: jest.fn(),
    }

    test('renders correctly', async () => {
      const wrapper = render(<FormInput {...props} />);
      expect(wrapper.baseElement).toMatchSnapshot();
    });

    test('renders correctly when there is validationError', async () => {
      const wrapper = render(<FormInput {...props} validationError={true} />);
      expect(wrapper.baseElement).toMatchSnapshot();
    });

    test(`should call setValue with ${props.value}`, async () => {
      render(<FormInput {...props} />);

      const input = screen.getByTestId(props.name);

      fireEvent.change(input, {target: { value: props.value}});
      expect(props.setValue).toHaveBeenCalledWith(props.value);
    });

    test(`when focused, should show label inside legend`, async () => {
      render(<FormInput {...props} />);

      fireEvent.focus(screen.getByTestId(props.name));
      const legend = screen.getByTestId(`legend-${PASSWORD_LABEL}`);

      expect(legend).toHaveTextContent(props.name);
    });

    test(`when out of focus and have value, show label inside legend`, async () => {
      render(<FormInput {...props} />);

      const input = screen.getByTestId(props.name);
      fireEvent.blur(input);

      const legend = screen.getByTestId(`legend-${PASSWORD_LABEL}`);
      expect(legend).toHaveTextContent(props.name);

    });

    test(`when out of focus without value, should not show label inside legend`, async () => {
      render(<FormInput {...props} value={''} />);

      const input = screen.getByTestId(props.name);
      fireEvent.blur(input);

      const legend = screen.getByTestId(`legend-${PASSWORD_LABEL}`);
      expect(legend).not.toHaveTextContent(props.name);

    });

    test('when clicked on eye icon, password input should have text type', async () => {
      render(<FormInput {...props} />);

      const eye = screen.getByTestId('eye');
      fireEvent.click(eye);

      const input = screen.getByTestId(props.name);

      expect(input.getAttribute('type')).toBe('text');
    });
  })

  describe('when input type is email', () => {
    const props = {
      type: 'email',
      name: 'email',
      value: 'randomemail@example.com',
      validationError: false,
      setValue: jest.fn(),
    }

    test('renders correctly', async () => {
      const wrapper = render(<FormInput {...props} />);
      expect(wrapper.baseElement).toMatchSnapshot();
    });

    test('renders correctly when there is validationError', async () => {
      const wrapper = render(<FormInput {...props} validationError={true} />);
      expect(wrapper.baseElement).toMatchSnapshot();
    });
  })

  describe('when input type is text', () => {
    const props = {
      type: 'text',
      name: 'Your Name',
      value: 'John Doe',
      validationError: false,
      setValue: jest.fn(),
    }

    test('renders correctly', async () => {
      const wrapper = render(<FormInput {...props} />);
      expect(wrapper.baseElement).toMatchSnapshot();
    });
  })

});