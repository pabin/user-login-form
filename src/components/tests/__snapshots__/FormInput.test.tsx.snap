// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormInput when input type is email renders correctly 1`] = `
<body>
  <div>
    <div
      class="container"
    >
      <fieldset
        class="fieldset"
      >
        <legend
          class="legend"
          data-testid="legend-email"
        >
          email
        </legend>
        <input
          class="input"
          data-testid="email"
          placeholder="email"
          type="email"
        />
      </fieldset>
    </div>
  </div>
</body>
`;

exports[`FormInput when input type is email renders correctly when there is validationError 1`] = `
<body>
  <div>
    <div
      class="container"
    >
      <fieldset
        class="fieldsetError fieldset"
      >
        <legend
          class="legendError legend"
          data-testid="legend-email"
        >
          email
        </legend>
        <input
          class="input"
          data-testid="email"
          placeholder="email"
          type="email"
        />
      </fieldset>
      <span
        class="helpText validationError"
        data-testid="emailError"
      >
        Please enter a valid email address
      </span>
    </div>
  </div>
</body>
`;

exports[`FormInput when input type is password renders correctly 1`] = `
<body>
  <div>
    <div
      class="container"
    >
      <fieldset
        class="fieldset"
      >
        <legend
          class="legend"
          data-testid="legend-Password"
        >
          Password
        </legend>
        <input
          class="input"
          data-testid="Password"
          placeholder="Password"
          type="password"
        />
        <img
          alt="eye"
          class="icon"
          data-testid="eye"
          height="18"
          src="eye.png"
          width="18"
        />
      </fieldset>
      <span
        class="helpText"
        data-testid="passwordHelpText"
      >
        Minimum 8 Characters
      </span>
    </div>
  </div>
</body>
`;

exports[`FormInput when input type is password renders correctly when there is validationError 1`] = `
<body>
  <div>
    <div
      class="container"
    >
      <fieldset
        class="fieldsetError fieldset"
      >
        <legend
          class="legendError legend"
          data-testid="legend-Password"
        >
          Password
        </legend>
        <input
          class="input"
          data-testid="Password"
          placeholder="Password"
          type="password"
        />
        <img
          alt="eye"
          class="icon"
          data-testid="eye"
          height="18"
          src="eye.png"
          width="18"
        />
      </fieldset>
      <span
        class="helpText validationError"
        data-testid="passwordHelpText"
      >
        Minimum 8 Characters
      </span>
    </div>
  </div>
</body>
`;

exports[`FormInput when input type is text renders correctly 1`] = `
<body>
  <div>
    <div
      class="container"
    >
      <fieldset
        class="fieldset"
      >
        <legend
          class="legend"
          data-testid="legend-Your Name"
        >
          Your Name
        </legend>
        <input
          class="input"
          data-testid="Your Name"
          placeholder="Your Name"
          type="text"
        />
      </fieldset>
    </div>
  </div>
</body>
`;
